### PR TITLE
[ShowHiddenChannels] Add last message info, fix collapsing

### PR DESCRIPTION
**Pull request by D3SOX**
- Show last message date
- Don't send messages API request
- Fix collapsing categories containing only hidden channels

(thx @CanadaHonk for the help)

### CODE REVIEW

File: src/plugins/showHiddenChannels.tsx [hunk 1]
```diff
@@ -89,13 +89,30 @@ export default definePlugin({
         channel._isHiddenChannel = !can(VIEW_CHANNEL, channel);
         return channel._isHiddenChannel;
     },
+    getDateFromSnowflake(snowflake: string) {
```
Comment (Vendicated) [inline src/plugins/showHiddenChannels.tsx:92]: Discord has a method for this
image

Comment (D3SOX) [reply src/plugins/showHiddenChannels.tsx:92]: Oh I didn't know that, replacing it...

**Comment by Vendicated** [comment]
could you please show a screenshot also?

**Comment by D3SOX** [comment]
> could you please show a screenshot also?

image

Not sure about the `Text` variants, I just copied the style from the topic info

File: src/plugins/showHiddenChannels.tsx (hunk 1 above)
Comment (D3SOX) [reply src/plugins/showHiddenChannels.tsx]: Done

**Review by Vendicated — APPROVED** [review]

**Comment by Vendicated** [comment]
you can add yourself in the authors section if you want

**Comment by D3SOX** [comment]
Ok pushed all my changes as of now, not sure how to handle the hardcoded variable names

**Comment by D3SOX** [comment]
Ok I noticed a major bug, pushing a fix soon

**Comment by Luna-devv** [comment]
Just as a small suggestion to this, what about a hotkey or smth similar so you can hide and show the hidden channels quickly if you, for example, stream discord or take screenshots

**Comment by MeguminSama** [comment]
Arguably it takes a few seconds to disable the plugin and CTRL+R

**Comment by Luna-devv** [comment]
I am in vc the whole time and wont rejoin and restart discord the whole time only for that lmao

**Comment by TymanWasTaken** [comment]
You could also show channel topic

File: src/plugins/showHiddenChannels.tsx [hunk 2]
```diff
@@ -55,27 +60,43 @@ export default definePlugin({
...
-                {
-                    match: /selectChannel:function\((\w)\){/g,
-                    replace: "selectChannel:function($1){if(Vencord.Plugins.plugins.ShowHiddenChannels.channelSelected($1)) return;"
-                }
-            ]
+            // Prevents Discord from trying to fetch message and create a 403 error
+            find: "fetchMessages:function",
+            replacement: {
+                match: /fetchMessages:function\((\w)\){/g,
+                replace: "fetchMessages:function($1){if(Vencord.Plugins.plugins.ShowHiddenChannels.isHiddenChannel($1)) return;"
```
Comment (Vendicated) [inline src/plugins/showHiddenChannels.tsx:79]: is there really no better way than this workaround? I posted a method that will likely work on Discord earlier

**Comment by D3SOX** [comment]
> You could also show channel topic

It's already there

File: src/plugins/showHiddenChannels.tsx (hunk 2 above)
Comment (D3SOX) [reply src/plugins/showHiddenChannels.tsx:79]: Where exactly? I haven't seen it

**Comment by Vendicated** [comment]
As for the toggling suggested by Luna, I think the best solution would be to add css classes to hidden channels and make a seperate css toggler plugin that can do it on a keybind. This way you can also hide other things

File: src/plugins/showHiddenChannels.tsx [hunk 3]
```diff
@@ -84,18 +105,33 @@ export default definePlugin({
...
+        if (!channel || channel.isDM() || channel.isGroupDM() || channel.isMultiUserDM())
+            return false;
+
+        // check for disallowed voice channels too so that they get hidden when collapsing the category
+        channel._isHiddenChannel = !can(VIEW_CHANNEL, channel) || (channel.type === 2 && !can(CONNECT, channel));
         return channel._isHiddenChannel;
     },
     channelSelected(channelData) {
+        if (!channelData) return false;
         const channel = ChannelStore.getChannel(channelData.channelId);
```
Comment (sneksnake) [inline src/plugins/showHiddenChannels.tsx:129]:
```suggestion
        if (!channel) return false;
```
discord crashes on startup for me without this
```
66716e85bf16d7f6a198.js:6052 TypeError: Cannot read property 'type' of undefined
    at Object.channelSelected (VencordRenderer:14)
    at vce (VM3171 WebpackModule583156:2)
    at i.handleHistoryChange (VM3171 WebpackModule583156:2)
    at e.t.addRouteChangeListener (66716e85bf16d7f6a198.js:4115)
    at o.i.componentDidMount (VM3171 WebpackModule583156:2)
    at Ss (WebpackModule364448:2)
    at Vu (WebpackModule364448:2)
    at t.unstable_runWithPriority (66716e85bf16d7f6a198.js:1663)
    at Ko (WebpackModule364448:2)
    at Bu (WebpackModule364448:2)
```

File: src/plugins/showHiddenChannels.tsx (hunk 2 above)
Comment (Vendicated) [reply src/plugins/showHiddenChannels.tsx:79]: https://discord.com/channels/1015060230222131221/1015063227299811479/1033709681241227295

Comment (D3SOX) [reply src/plugins/showHiddenChannels.tsx:79]: Thx, will adjust when I'm back home :smiley:

**Review by Vendicated — APPROVED** [review]

**Pull request merged**